test/integration/logic/resolvable/UnifyRelationConcludableTest.java

```diff
@@ -787,6 +856,314 @@ public void relation_more_players_than_rule_relation_fails_unify() {
+        unifiable(conclusions.get(2), "{$p isa student;$q isa student-driver;}", parents, Lists.newArrayList(0, 1));
+
+        unifiable(conclusions.get(3), "{$p isa driving-hire; $q isa student-driver;}", parents,  new ArrayList<>());
+        unifiable(conclusions.get(3), "{$p isa driving-hire; $q isa driving-hire;}", parents, Lists.newArrayList(1, 2));
+        unifiable(conclusions.get(3), "{$p isa driving-hire; $q isa organisation;}", parents, Lists.newArrayList(1, 2));
+    }
 
-    // TODO: rule unification pruning tests based based on types
+    @Test
+    public void differentIllegalRelationVariants() {
+        String student = Bytes.bytesToHexString(instanceOf("student").getIID());
+        String organisation = Bytes.bytesToHexString(instanceOf("organisation").getIID());
+        List<String> parents = Lists.newArrayList(
+                String.format("{(taxi: $x, night-shift-driver: $y); $x iid %s;}", organisation),
+                String.format("{(taxi: $x, night-shift-driver: $y); $x iid %s;}", student),
+                String.format("{(employer: $x, night-shift-driver: $y); $y iid %s;}", organisation),
+                String.format("{(employer: $x, night-shift-driver: $y); $y iid %s;}", student)
+        );
+
+        List<String> conclusions = Lists.newArrayList(
+                "(employer: $p, employee: $q) isa employment",
+                "(part-time-employer: $p, part-time-employee: $q) isa part-time-employment",
+                "(taxi: $p, day-shift-driver: $q) isa part-time-driving"
+        );
+        conclusions.forEach(conclusion -> unifiable(conclusion,"{$p isa driving-hire; $q isa student-driver;}", parents, new ArrayList<>()));
+    }
+
+    private FunctionalIterator<Unifier> unifiers(String parent, String ruleConclusion, String rulePremises){
+        Conjunction parentConjunction = resolvedConjunction(parent, logicMgr);
+        Concludable.Relation queryConcludable = Concludable.create(parentConjunction).stream()
+                .filter(Concludable::isRelation)
+                .map(Concludable::asRelation)
+                .findFirst().orElse(null);
+        Rule rule = createRule("test-rule", rulePremises, ruleConclusion, logicMgr);
+        return queryConcludable.unify(rule.conclusion(), conceptMgr);
+    }
+
+    private void nonExistentUnifier(String parent, String ruleConclusion, String rulePremises){
+        assertFalse(unifiers(parent, ruleConclusion, rulePremises).hasNext());
+    }
+
+    private Unifier uniqueUnifier(String parent, String ruleConclusion, String rulePremises){
+        List<Unifier> unifiers = unifiers(parent, ruleConclusion, rulePremises).toList();
+        assertEquals(1, unifiers.size());
+        return unifiers.iterator().next();
+    }
+
+    private void unifiable(String ruleConclusion, String rulePremises, List<String> parents, List<Integer> unifiableParents){
```

Comment:
since this is specific to testing unifications within a list, and also does verification similarly to the next method, we should name them similarly:
`verify...`
maybe
`verifyUnificationsOnlySucceedFor(ruleConclusion, ruleCondition, parentRelations, successfulParentIndices)`